<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод:</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

In [1]:
import pandas as pd
import numpy as np

from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score


In [2]:
try:
    data = pd.read_csv("/datasets/insurance.csv")
except:
    data = pd.read_csv("datasets/insurance.csv")

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [4]:
data.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [5]:
data.duplicated().sum()

153

In [6]:
data[data.duplicated()].head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
281    1     39.0   48100.0            1                  0
488    1     24.0   32900.0            1                  0
513    0     31.0   37400.0            2                  0
718    1     22.0   32600.0            1                  0
785    0     20.0   35800.0            0                  0

In [7]:
data.isnull().sum()

Пол                  0
Возраст              0
Зарплата             0
Члены семьи          0
Страховые выплаты    0
dtype: int64

In [8]:
data.drop_duplicates(inplace=True)

In [9]:
data.duplicated().sum()

0

In [10]:
data["Возраст"] = data["Возраст"].astype("int")

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4847 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                4847 non-null   int64  
 1   Возраст            4847 non-null   int64  
 2   Зарплата           4847 non-null   float64
 3   Члены семьи        4847 non-null   int64  
 4   Страховые выплаты  4847 non-null   int64  
dtypes: float64(1), int64(4)
memory usage: 227.2 KB


### Вывод:

* Все данные в датасете полностью заполнены. Пропусков нет.
* Выявлены и удалены 153 дубликата.
* Типы данных в столбце `Возраст` изменен на `int`.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** При умножение признаки на обратимую матрицу - качество линейной регресии не изменится.

**Обоснование:** 

Примем: 
$$
a' = X'w
$$

$$
X' = XZ
$$

где, Z - Обратная матрица


Свойства, которые будут использоваться:

$$
AA^{-1} = A^{-1}A = E
$$

$$
AE = EA = A
$$

$$
(AB)^{-1} =B^{-1} A^{-1}
$$

$$
(AB)^{T} = B^{T} A^{T}
$$

Подставим в a' формулу обучения:

$$
a' = X'(X'^{T}X)^{-1}X'^{T}y
$$

Подставим X':

$$
a' = XZ((XZ)^{T}XZ)^{-1}Z^{T}X^{T}y
$$

Раскроем скобки с помощью свойств матриц:

$$
a' = XZ(Z^{T}X^{T}XZ)^{-1}Z^{T}X^{T}y = XZZ^{-1}(Z^{T}X^{T}X)^{-1}Z^{T}X^{T}y
$$

$$
ZZ^{-1} = E 
$$
$$
(Z^{T})^{-1}Z^{T} = E
$$

В итоге получается:

$$
a' = XE(Z^{T}X^{T}X)^{-1}Z^{T}X^{T}y = X(X^{T}X)^{-1}EX^{T}y = X(X^{T}X)^{-1}X^{T}y = Xw
$$

$$
a = a'
$$

Таким образом, получается формулa, что и была в начале. Можно сделать вывод, что домножение признаков на обратимую матрицу не влияет на предсказание модели.

## Алгоритм преобразования

**Алгоритм**

1. Обучить модель без умножения на обратную матрицу.
2. Получить значение R2
3. Создать рандомную обратную матрицу и умножить ее на матрицу признаков
4. Обучить модель
5. Получить значение R2
6. Сравнить R2. Они должны быть одинаковы. Если нет, то где-то ошибка

**Обоснование**

В прошлому пункте было доказано, что a = a'

## Проверка алгоритма

In [12]:
features = data.drop(['Страховые выплаты'], axis = 1)
target = data['Страховые выплаты']

In [13]:
model =  LinearRegression()
model.fit(features,target)
prediction = model.predict(features)
r2_first = r2_score(target,prediction).round(9)

r2_first

0.430201004

Для существования обратной матрицы необходимо и достаточно, чтобы матрица  была невырожденной, то есть, чтобы ее определитель не был равен нулю. Поэтому зададим определитель равный 0 и используем while. Как только определитель перестанет быть равным 0 - выход из цикла.

In [14]:
def matrica(i,j):
    opr = 0 
    matrica = 0
    while opr == 0:
        matrica = np.random.normal(size=(i, j))
        opr = np.linalg.det(matrica)
    return matrica

In [15]:
M = matrica(4,4)
M

array([[ 0.64687708,  0.96179071,  0.72234788, -0.83917259],
       [-0.34895054, -0.53073765,  1.07992796,  0.52360981],
       [-1.94787615,  0.95348648,  0.36564884, -0.25092136],
       [ 0.85828368,  0.2487998 , -0.80559439,  0.03153046]])

In [16]:
features_new = features @ M

In [17]:
model =  LinearRegression()
model.fit(features_new,target)
prediction = model.predict(features_new)
r2_second = r2_score(target,prediction).round(9)

r2_second

0.430201004

In [18]:
if r2_first == r2_second:
    print("Метрики R2 обеих моделей одиннаковые")
else:
    print("Ошибка")

Метрики R2 обеих моделей одиннаковые


In [19]:
result = pd.DataFrame(data= [r2_first,r2_second], 
                     columns=['R2'], 
                     index=['Линейная регрессия',
                            'Линейная регрессия на преобразованных признаках'])

result

R2
Линейная регрессия                               0.430201
Линейная регрессия на преобразованных признаках  0.430201

### Вывод:

Метрики R2 обеих моделей одиннаковые, следовательно, умножение исходных данных на случайную обратимую матрицу может быть применено для их шифрования.

## Вывод:

1. Были изучены полученные данные. В результате выявлены дубликаты. Пропусков не обнаружено;
2. Удалены дубликаты и изменен тип данных в столбце `Возраст` на `int`;
3. Составлен алгоритм преобразования и доказано, что домножение признаков на обратимую матрицу не влияет на предсказание модели;
4. Проверен алгоритм. В итоге метрики R2 до и после преобразовния получились одинаковые. Следовательно, умножение исходных данных на случайную обратимую матрицу может быть применено для шифрования.